File: performance/convert_perf_timestamp.cpp
#include <chrono>
#include <iostream>
#include <iomanip>

int main(int argc, char* argv[]) {
  // const auto start_time = std::chrono::steady_clock::now().time_since_epoch();
  // std::cout << start_time.count() << "\n";

  // if (argc < 2) { return 1; }
  // const auto start_time = std::chrono::steady_clock::now().time_since_epoch();
  // const auto real_time_stamp = std::chrono::system_clock::now();
  // const auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(start_time);
  // long input_ms = std::stod(argv[1]) * 1000.0;
  // const auto diff_ms = input_ms - ms.count();
  // const auto result_time_stamp = real_time_stamp + std::chrono::milliseconds(diff_ms);
  // std::cout << "ms: " << ms.count() << '\n';
  // auto time_t_result = std::chrono::system_clock::to_time_t(result_time_stamp);
  // std::tm* tm_result = std::localtime(&time_t_result);
  // std::cout << "Date: " << std::put_time(tm_result, "%Y-%m-%d") << '\n';
  // std::cout << "Time: " << std::put_time(tm_result, "%H:%M:%S") << '\n';

  if (argc < 4) { return 1; }
  long input_real_time_ns = std::stoul(argv[1]);
  const auto real_time_stamp = std::chrono::system_clock::time_point(std::chrono::nanoseconds(input_real_time_ns));
  {
    auto time_t_result = std::chrono::system_clock::to_time_t(real_time_stamp);
    std::tm* tm_result = std::localtime(&time_t_result);
    std::cout << "Date: " << std::put_time(tm_result, "%Y-%m-%d") << '\n';
    std::cout << "Time: " << std::put_time(tm_result, "%H:%M:%S") << '\n';
  }
  long ms = std::stoul(argv[2]) / 1e6;
  long input_ms = std::stod(argv[3]) * 1e3;
  const auto diff_ms = input_ms - ms;
  std::cout << "diff_ms: " << diff_ms << '\n';
  const auto result_time_stamp = real_time_stamp + std::chrono::milliseconds(diff_ms);
  std::cout << "ms: " << ms << '\n';
  {
    auto time_t_result = std::chrono::system_clock::to_time_t(result_time_stamp);
    std::tm* tm_result = std::localtime(&time_t_result);
    std::cout << "Date: " << std::put_time(tm_result, "%Y-%m-%d") << '\n';
    std::cout << "Time: " << std::put_time(tm_result, "%H:%M:%S") << '\n';
  }
  return 0;
}

File: performance/convert_perf_timestamp.py
import sys
import datetime

def main():
    if len(sys.argv) < 4:
        return 1
    
    # 获取命令行参数
    input_real_time_ns = int(sys.argv[1])
    
    # 将纳秒时间戳转换为datetime对象
    real_time_stamp = datetime.datetime.fromtimestamp(input_real_time_ns / 1e9)
    
    # 输出原始时间戳的日期和时间
    print(f"Date: {real_time_stamp.strftime('%Y-%m-%d')}")
    print(f"Time: {real_time_stamp.strftime('%H:%M:%S')}")
    
    # 计算时间差
    ms = int(sys.argv[2]) / 1e6
    input_ms = float(sys.argv[3]) * 1e3
    diff_ms = input_ms - ms
    
    print(f"diff_ms: {int(diff_ms)}")
    
    # 计算结果时间戳
    result_time_stamp = real_time_stamp + datetime.timedelta(milliseconds=diff_ms)
    
    print(f"ms: {int(ms)}")
    
    # 输出结果时间戳的日期和时间
    print(f"Date: {result_time_stamp.strftime('%Y-%m-%d')}")
    print(f"Time: {result_time_stamp.strftime('%H:%M:%S')}")
    
    return 0

if __name__ == "__main__":
    main()
